scripts/bert/finetune_squad.py

```diff
@@ -0,0 +1,485 @@
+console.setLevel(logging.INFO)
+console.setFormatter(formatter)
+log.addHandler(console)
+log.addHandler(fh)
+
+log.info(args)
+
+model_name = args.bert_model
+dataset_name = args.bert_dataset
+only_predict = args.only_predict
+model_parameters = args.model_parameters
+lower = args.uncased
+
+epochs = args.epochs
+batch_size = args.batch_size
+test_batch_size = args.test_batch_size
+lr = args.lr
+ctx = mx.cpu() if not args.gpu else mx.gpu()
+
+accumulate = args.accumulate
+log_interval = args.log_interval * accumulate if accumulate else args.log_interval
+if accumulate:
+    log.info('Using gradient accumulation. Effective batch size = {}'.
+             format(accumulate*batch_size))
+
+optimizer = args.optimizer
+warmup_ratio = args.warmup_ratio
+
+
+version_2 = args.version_2
+null_score_diff_threshold = args.null_score_diff_threshold
+
+max_seq_length = args.max_seq_length
+doc_stride = args.doc_stride
+max_query_length = args.max_query_length
+n_best_size = args.n_best_size
+max_answer_length = args.max_answer_length
+
+if max_seq_length <= max_query_length + 3:
+    raise ValueError('The max_seq_length (%d) must be greater than max_query_length '
+                     '(%d) + 3' % (max_seq_length, max_query_length))
+
+bert, vocab = nlp.model.get_model(
+    name=model_name,
+    dataset_name=dataset_name,
+    pretrained=not model_parameters,
+    ctx=ctx,
+    use_pooler=False,
+    use_decoder=False,
+    use_classifier=False)
+
+batchify_fn = nlp.data.batchify.Tuple(
+    nlp.data.batchify.Stack(),
+    nlp.data.batchify.Pad(axis=0, pad_val=vocab[vocab.padding_token]),
+    nlp.data.batchify.Pad(axis=0, pad_val=vocab[vocab.padding_token]),
+    nlp.data.batchify.Stack('float32'),
+    nlp.data.batchify.Stack('float32'),
+    nlp.data.batchify.Stack('float32'))
+
+berttoken = nlp.data.BERTTokenizer(vocab=vocab, lower=lower)
+
+net = BertForQA(bert=bert)
+if not model_parameters:
+    net.span_classifier.initialize(init=mx.init.Normal(0.02), ctx=ctx)
+else:
+    net.load_parameters(model_parameters, ctx=ctx)
+net.hybridize(static_alloc=True)
+
+loss_function = BertForQALoss()
+loss_function.hybridize(static_alloc=True)
+
+
+def train():
+    """Training function."""
+    log.info('Loader Train data...')
+    if version_2:
+        train_data = SQuAD('train', version='2.0')
+    else:
+        train_data = SQuAD('train', version='1.1')
+    log.info('Number of records in Train data:{}'.format(len(train_data)))
+
+    train_data_transform = preprocess_dataset(train_data, SQuADTransform(
+        berttoken,
+        max_seq_length=max_seq_length,
+        doc_stride=doc_stride,
+        max_query_length=max_query_length,
+        is_pad=True,
+        is_training=True))
+    log.info('The number of examples after preprocessing:{}'.format(
+        len(train_data_transform)))
+
+    train_dataloader = mx.gluon.data.DataLoader(
+        train_data_transform, batch_size=batch_size,
+        batchify_fn=batchify_fn, num_workers=4, shuffle=True)
```

Comment:
#590 . Use bucketing here to improve performance, such as [nlp.data.sampler.FixedBucketSampler](https://gluon-nlp.mxnet.io/api/notes/data_api.html#bucketing-and-dataloader).

Reply:
Agree, use `batch_sampler=FixedBucketSampler` could be helpful.

Reply:
After using `FixedBucketSampler`, GPU memory will also increase, and will gradually increase to the same size as before.